File: cli/src/core/NotePromptFormatter.ts
/**
 * Note prompt formatter — renders active notes as the <notes> XML block for SUMMARIZE.
 *
 * Reads each note's body bytes from `entry.sourcePath` (both `snippet` and
 * `markdown` formats — the format field only differentiates the on-disk origin
 * and panel icon; the prompt-side read path is identical). Body lengths are
 * smaller than plans (notes are typically <2KB), so the per-note and total
 * caps are tighter.
 */

import { readFile } from "node:fs/promises";
import { createLogger } from "../Logger.js";
import type { NoteEntry } from "../Types.js";
import { escapeForAttr, escapeForText } from "./PromptXmlEscape.js";

const log = createLogger("NotePromptFormatter");

const DEFAULT_MAX_CHARS_PER_NOTE = 4000;
const DEFAULT_MAX_TOTAL_CHARS = 12000;

export interface FormatNotesOptions {
	readonly maxCharsPerNote?: number;
	readonly maxTotalChars?: number;
}

export async function formatNotesBlock(
	entries: ReadonlyArray<NoteEntry>,
	opts: FormatNotesOptions = {},
): Promise<string> {
	if (entries.length === 0) return "";

	const maxPerNote = opts.maxCharsPerNote ?? DEFAULT_MAX_CHARS_PER_NOTE;
	const maxTotal = opts.maxTotalChars ?? DEFAULT_MAX_TOTAL_CHARS;

	// Respect the caller's order (relevance-ranked, most relevant first) so
	// over-budget truncation drops the least relevant, not the oldest.
	const ordered = entries;

	const selected: Array<{ entry: NoteEntry; body: string }> = [];
	let totalLen = 0;
	for (const entry of ordered) {
		const body = await readNoteBody(entry);
		const rendered = renderOneNote(entry, body, maxPerNote);
		if (totalLen + rendered.length > maxTotal) break;
		selected.push({ entry, body });
		totalLen += rendered.length;
	}

	if (selected.length === 0) return "";

	const inner = selected.map(({ entry, body }) => renderOneNote(entry, body, maxPerNote)).join("\n");
	return `<notes>\n${inner}\n</notes>`;
}

async function readNoteBody(entry: NoteEntry): Promise<string> {
	if (!entry.sourcePath) return "";
	try {
		return await readFile(entry.sourcePath, "utf-8");
	} catch (err) {
		// log.warn (not debug): see the parallel comment in
		// PlanPromptFormatter.readPlanBody — empty <note> body in the
		// SUMMARIZE prompt is indistinguishable from a genuinely-empty note
		// unless this read failure leaves a trace in debug.log.
		log.warn("Cannot read note markdown %s: %s", entry.sourcePath, (err as Error).message);
		return "";
	}
}

function renderOneNote(entry: NoteEntry, body: string, maxChars: number): string {
	const truncated =
		body.length > maxChars
			? `${body.slice(0, maxChars)}\n…[truncated, ${body.length - maxChars} more chars]`
			: body;
	const lines: string[] = [
		`<note id="${escapeForAttr(entry.id)}" format="${escapeForAttr(entry.format)}">`,
		`  <title>${escapeForText(entry.title)}</title>`,
		`  <updated-at>${escapeForAttr(entry.updatedAt)}</updated-at>`,
	];
	if (truncated.length > 0) {
		lines.push("  <content>");
		lines.push(escapeForText(truncated));
		lines.push("  </content>");
	}
	lines.push("</note>");
	return lines.join("\n");
}

File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/NotePromptFormatter.kt
package ai.jolli.jollimemory.core

import ai.jolli.jollimemory.core.references.PromptRenderer
import java.io.File

/**
 * NotePromptFormatter — renders active note entries into a `<notes>` XML block for
 * the summary prompt. Kotlin port of the CLI `NotePromptFormatter.ts`.
 *
 * Ordering: the caller's order is preserved verbatim (relevance-ranked, most
 * relevant first) so over-budget truncation drops the least relevant, not the
 * oldest. An unranked caller gets insertion order.
 */
object NotePromptFormatter {

    private val log = JmLogger.create("NotePromptFormatter")

    private const val DEFAULT_MAX_CHARS_PER_NOTE = 4000
    private const val DEFAULT_MAX_TOTAL_CHARS = 30000

    /** Reads the note body from its sourcePath; empty string when missing/unreadable. */
    private fun readNoteBody(entry: NoteEntry): String {
        val path = entry.sourcePath ?: return ""
        return try {
            val f = File(path)
            if (f.exists()) f.readText(Charsets.UTF_8) else ""
        } catch (_: Exception) {
            ""
        }
    }

    private fun truncate(s: String, max: Int): String {
        if (s.length <= max) return s
        return "${s.take(max)}\n…[truncated, ${s.length - max} more chars]"
    }

    /** Renders one note as a `<note>` element with its (truncated) body. */
    private fun renderOneNote(entry: NoteEntry, body: String, maxPerNote: Int): String {
        val lines = mutableListOf("<note id=\"${PromptRenderer.escapeForAttr(entry.id)}\" title=\"${PromptRenderer.escapeForAttr(entry.title)}\">")
        val trimmed = body.trim()
        if (trimmed.isNotEmpty()) {
            lines.add(PromptRenderer.escapeForText(truncate(trimmed, maxPerNote)))
        }
        lines.add("</note>")
        return lines.joinToString("\n")
    }

    /**
     * Renders the `<notes>` block for [entries], in caller order, within budget.
     * Returns "" when there is nothing to render.
     */
    fun formatNotesBlock(
        entries: List<NoteEntry>,
        maxCharsPerNote: Int = DEFAULT_MAX_CHARS_PER_NOTE,
        maxTotalChars: Int = DEFAULT_MAX_TOTAL_CHARS,
    ): String {
        if (entries.isEmpty()) return ""

        val selected = mutableListOf<Pair<NoteEntry, String>>()
        var totalLen = 0
        for (entry in entries) {
            val body = readNoteBody(entry)
            val rendered = renderOneNote(entry, body, maxCharsPerNote)
            if (totalLen + rendered.length > maxTotalChars) break
            selected.add(entry to body)
            totalLen += rendered.length
        }
        if (selected.isEmpty()) return ""

        val inner = selected.joinToString("\n") { (entry, body) -> renderOneNote(entry, body, maxCharsPerNote) }
        log.info("Formatted notes block: %d of %d note(s), %d chars", selected.size, entries.size, inner.length)
        return "<notes>\n$inner\n</notes>"
    }
}
